provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+import { EventEmitter } from 'events';
+import { errors, X509, SharedAccessSignature } from 'azure-iot-common';
+import * as machina from 'machina';
+import * as Provisioning from './transport_interface';
+import * as dbg from 'debug';
+const debug = dbg('azure-device-provisioning:transport-fsm');
+
+export interface TransportHandlers {
+  connect(callback: (err?: Error) => void): void;
+  disconnect(callback: (err?: Error) => void): void;
+  registrationRequest(registrationId: string, authorization: SharedAccessSignature | X509 | string, requestBody: any, forceRegistration: boolean, callback: (err?: Error, responseBody?: any, result?: any, pollingInterval?: number) => void): void;
+  queryOperationStatus(registrationId: string, operationId: string, callback: (err?: Error, responseBody?: any, result?: any, pollingInterval?: number) => void): void;
+  getErrorResult(result: any): any;
+}
+
+export class TransportStateMachine extends EventEmitter implements Provisioning.Transport {
+  private _fsm: machina.Fsm;
+  private _pollingTimer: any;
+  private _transport: TransportHandlers;
+
+  /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_001: [ The `constructor` shall accept no arguments ] */
+  constructor() {
+    super();
+  }
+
+  initialize(transport: TransportHandlers): void {
+    this._transport = transport;
+
+    this._fsm = new machina.Fsm({
+      namespace: 'provisioning-transport',
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err) => {
+            this._pollingTimer = null;
+            if (callback) {
+              callback(err);
+            }
+          },
+          connect: (callback) => {
+            this._fsm.transition('connecting', callback);
+          },
+          register: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_010: [ `register` shall connect the transport if it is not connected. ] */
+            this._fsm.handle('connect', (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.transition('sendingRegistrationReqest', callback, registrationId, authorization, requestBody, forceRegistration);
+              }
+            });
+          },
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_027: [ If a registration is in progress, `disconnect` shall cause that registration to fail with an `OperationCancelledError`. ] */
+          responseReceived: (callback) => callback(new errors.OperationCancelledError()),
```

Comment:
are we going to get there? receive a response after being disconnected?

Reply:
wondering if ew need a "pending operation queue" instead that you can flush and call the callbacks with an error

---
In reply to: [147223085](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r147223085) [](ancestors = 147223085)